text/0000-osstring-string-interface.md

```diff
@@ -0,0 +1,239 @@
+
+```rust
+let mut opts = &OsString::from("vL/path")[..]; // Leading '-' has already been removed
+while let Some((ch, rest)) = opts.slice_shift_char() {
+    opts = rest;
+    match ch {
+        'v' => { verbose = true; }
+        'L' => { /* interpret remainder as a link path */ }
+        ....
+    }
+}
+```
+
+### `split_off_str`
+
+This is intended for interpreting "tagged" OS strings, for example
+rustc's `-L [KIND=]PATH` arguments.  It is expected that such tags
+will usually be UTF-8.  Example:
+```rust
+let s = OsString::from("dylib=/path");
+
+let (name, kind) = match s.split_off_str('=') {
+    None => (&*s, cstore::NativeUnknown),
+    Some(("dylib", name)) => (name, cstore::NativeUnknown),
+    Some(("framework", name)) => (name, cstore::NativeFramework),
+    Some(("static", name)) => (name, cstore::NativeStatic),
+    Some((s, _)) => { error(...) }
+};
+```
+
+### `split`
+
+This is similar to the similarly named function on `str`, except the
+splitting boundary is restricted to be an ASCII character instead of a
+general pattern.  ASCII characters have well-defined meanings in both
+flavors of OS string, and the portions before and after such a
+character are always well-formed OS strings.
+
+This is intended for interpreting OS strings containing several paths.
+Using this function will generally restrict the allowed paths to those
+not containing the separator, but this is a common limitation already
+in such interfaces.  For example, rustc's `--emit dep-info=bar.d,link`
+could be processed as:
+```rust
+let arg = OsString::from("dep-info=bar.d,link");
+
+for part in arg.split(',') {
+    match part.split_off_str('=') {
+        ...
+    }
+}
+```
+
+## `SliceConcatExt`
+
+Implement the trait
+```rust
+impl<S> SliceConcatExt<OsStr> for [S] where S: Borrow<OsStr> {
+    type Output = OsString;
+    ...
+}
+```
+
+This has the same behavior as the `str` version, except that it works
+on OS strings.  It is intended as a more convenient and efficient way
+of building up an `OsString` from parts than repeatedly calling
+`push`.
+
+# Drawbacks
+
+This is a somewhat unusual string interface in that much of the
+functionality either accepts or returns a different type of string
+than the one the interface is designed to work with (`str` instead of
+the probably expected `OsStr`).
+
+# Alternatives
+
+## Interfaces without `str`
+
+Versions of the `*_str` functions that take or return `&OsStr`s seem
+more natural, but in at least some of the cases it is not possible to
+implement such an interface.  For example, on Windows, the following
+should hold using a hypothetical `remove_prefix(&self, &OsStr) ->
+Option<&OsStr>`:
+
+```rust
+let string = OsString::from("😺"); // [0xD83D, 0xDE3A] in UTF-16
+let prefix: OsString = OsStringExt::from_wide(&[0xD83D]);
+let suffix: OsString = OsStringExt::from_wide(&[0xDE3A]);
+
+assert_eq!(string.remove_prefix(&prefix[..]), Some(&suffix[..]));
```

Comment:
It is possible to write a `starts_with` variant that does this (and one will be included in the next version of this RFC), but for `remove_prefix` you also have to construct an `&OsStr` representing the rest of the string, which is a problem if that `OsStr` should start with a trail surrogate that is combined with something in the in-memory representation.  (I suppose it could return a `Cow<OsStr>`, but I'm not planning on proposing that unless someone else thinks it's a really good idea.)